### PR TITLE
AP_MSP: MSP_RAW_GPS cog should be decidegrees not centidegrees

### PR DESCRIPTION
**Pull request by catkira**
image

source: http://www.multiwii.com/wiki/index.php?title=Multiwii_Serial_Protocol

### CODE REVIEW

**Comment by catkira** [comment]
I think we should also make ground_course_dd uint16_t

**Comment by peterbarker** [comment]
Looks likely. 

Commit message should be `AP_MSP:`, not `fix:`.  Want to fix that or shall I?

**Comment by catkira** [comment]
I can fix. Should I also change it to uint16_t?

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp [hunk 1]
```diff
@@ -217,7 +217,7 @@ void AP_MSP_Telem_Backend::update_gps_state(gps_state_t &gps_state)
         gps_state.lon = loc.lng;
         gps_state.alt_m = loc.alt/100; // 1m resolution
         gps_state.speed_cms = gps.ground_speed() * 100;
-        gps_state.ground_course_cd = gps.ground_course_cd();
+        gps_state.ground_course_dd = gps.ground_course_cd() * 10;
```
Comment (amilcarlucas) [inline libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:220]: To convert from centi-degrees to deci-degrees you need to divide by 10

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp (hunk 1 above)
Comment (catkira) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:220]: yes, good catch :)

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp [hunk 2]
```diff
@@ -217,7 +217,7 @@ void AP_MSP_Telem_Backend::update_gps_state(gps_state_t &gps_state)
         gps_state.lon = loc.lng;
         gps_state.alt_m = loc.alt/100; // 1m resolution
         gps_state.speed_cms = gps.ground_speed() * 100;
-        gps_state.ground_course_cd = gps.ground_course_cd();
+        gps_state.ground_course_dd = gps.ground_course_cd() / 10;
```
Comment (peterbarker) [inline libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:220]:
```suggestion
        gps_state.ground_course_dd = gps.ground_course_cd() * 0.1;
```

sorry, its an efficiency thing; multiplication is much faster than division.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_MSP/AP_MSP_Telem_Backend.cpp (hunk 2 above)
Comment (nexton-winjeel) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:220]: Two things:
1. This is integer division. Adding a `float` conversion is *not* efficient.
2. If this level of micro-optimisation matters so much, shouldn't we just enable the `-freciprocal-math` compiler optimisation? (Compiler Explorer example: https://godbolt.org/z/5zn7ejnxK)

Comment (peterbarker) [reply libraries/AP_MSP/AP_MSP_Telem_Backend.cpp:220]: >     1. This is integer division. Adding a `float` conversion is _not_ efficient.

Yep, my bad.  Reviewing too much floating point stuff moving from integer to floating point recently.

>     2. If this level of micro-optimisation matters so much, shouldn't we just enable the `-freciprocal-math` compiler optimisation? (Compiler Explorer example: https://godbolt.org/z/5zn7ejnxK)

I know someone looked at that at some stage but can't remember the resolution.

**Comment by catkira** [comment]
@rmackay9 can you backport this to plane 4.5.6 ?

**Pull request merged**